File: scripts/daily/generate-daily-brief.js
#!/usr/bin/env node

/**
 * Daily Brief Generator
 * Orchestrates data collection and generates the daily intelligence brief
 */

const fs = require('fs');
const path = require('path');

// Import data collectors
const { fetchCurrentWeather, fetchForecast, formatWeatherMarkdown } = require('./lib/weather.js');
const { fetchNews, formatNewsMarkdown } = require('./lib/news.js');
const { fetchSpaceWeatherAlerts, fetchKpIndex, formatSpaceWeatherMarkdown } = require('./lib/space-weather.js');
const { fetchQuote, formatQuoteMarkdown } = require('./lib/quote.js');
const { fetchTrendingRepos, formatTrendingMarkdown } = require('./lib/github-trending.js');

// Configuration
const README_PATH = path.join(__dirname, '..', '..', 'README.md');
const DAILY_DIR = path.join(__dirname, '..', '..', 'daily');
const ASSETS_DIR = path.join(__dirname, '..', '..', 'assets', 'daily');

/**
 * Ensure directories exist
 */
function ensureDirectories() {
  if (!fs.existsSync(DAILY_DIR)) {
    fs.mkdirSync(DAILY_DIR, { recursive: true });
  }
  if (!fs.existsSync(ASSETS_DIR)) {
    fs.mkdirSync(ASSETS_DIR, { recursive: true });
  }
}

/**
 * Get current date in various formats
 */
function getDateFormats() {
  const now = new Date();
  return {
    iso: now.toISOString().split('T')[0], // YYYY-MM-DD
    long: now.toLocaleDateString('en-US', { 
      weekday: 'long', 
      year: 'numeric', 
      month: 'long', 
      day: 'numeric' 
    }),
    time: now.toLocaleTimeString('en-US', { 
      hour: '2-digit', 
      minute: '2-digit',
      timeZoneName: 'short'
    })
  };
}

/**
 * Fetch all data with error handling
 */
async function fetchAllData() {
  const apiKey = process.env.OPENWEATHER_API_KEY;
  const ghToken = process.env.GITHUB_TOKEN || process.env.GH_TOKEN;
  
  console.log('🔄 Fetching data from all sources...\n');
  
  const results = {
    weather: null,
    forecast: null,
    news: null,
    spaceAlerts: null,
    kpIndex: null,
    quote: null,
    trending: null
  };
  
  // Fetch all data in parallel with individual error handling
  try {
    const [weather, forecast, news, spaceAlerts, kpIndex, quote, trending] = await Promise.allSettled([
      fetchCurrentWeather(apiKey),
      fetchForecast(apiKey),
      fetchNews(5),
      fetchSpaceWeatherAlerts(),
      fetchKpIndex(),
      fetchQuote(),
      fetchTrendingRepos(ghToken)
    ]);
    
    results.weather = weather.status === 'fulfilled' ? weather.value : null;
    results.forecast = forecast.status === 'fulfilled' ? forecast.value : null;
    results.news = news.status === 'fulfilled' ? news.value : null;
    results.spaceAlerts = spaceAlerts.status === 'fulfilled' ? spaceAlerts.value : null;
    results.kpIndex = kpIndex.status === 'fulfilled' ? kpIndex.value : null;
    results.quote = quote.status === 'fulfilled' ? quote.value : null;
    results.trending = trending.status === 'fulfilled' ? trending.value : null;
    
    // Log what succeeded
    console.log('✅ Weather:', results.weather ? 'OK' : 'Failed');
    console.log('✅ Forecast:', results.forecast ? 'OK' : 'Failed');
    console.log('✅ News:', results.news ? `${results.news.length} stories` : 'Failed');
    console.log('✅ Space Weather:', results.spaceAlerts ? 'OK' : 'Failed');
    console.log('✅ Quote:', results.quote ? 'OK' : 'Failed');
    console.log('✅ Trending Repos:', results.trending ? `${results.trending.length} repos` : 'Failed');
    console.log('');
    
  } catch (error) {
    console.error('⚠️  Error during data fetch:', error.message);
  }
  
  return results;
}

/**
 * Generate markdown content for the daily brief
 */
function generateBriefMarkdown(data, dates) {
  let content = '';
  
  // Header with date
  content += `<div align="center">\n\n`;
  content += `# 📅 Daily Brief\n\n`;
  content += `**${dates.long}**\n\n`;
  content += `</div>\n\n`;
  content += `---\n\n`;
  
  // Quote
  if (data.quote) {
    content += `<details>\n<summary><b>💭 Quote of the Day</b></summary>\n\n`;
    content += formatQuoteMarkdown(data.quote);
    content += `\n</details>\n\n`;
  }
  
  // Weather
  content += `<details>\n<summary><b>🌤️ Weather Report</b></summary>\n\n`;
  content += formatWeatherMarkdown(data.weather, data.forecast);
  content += `\n</details>\n\n`;
  
  // News
  content += `<details>\n<summary><b>📰 Global Intelligence News</b></summary>\n\n`;
  content += formatNewsMarkdown(data.news);
  content += `\n</details>\n\n`;
  
  // Space Weather
  content += `<details>\n<summary><b>🌌 Space Weather</b></summary>\n\n`;
  content += formatSpaceWeatherMarkdown(data.spaceAlerts, data.kpIndex);
  content += `\n</details>\n\n`;
  
  // GitHub Trending
  content += `<details>\n<summary><b>🔥 Trending on GitHub</b></summary>\n\n`;
  content += formatTrendingMarkdown(data.trending);
  content += `\n</details>\n\n`;
  
  content += `---\n\n`;
  content += `<div align="center">\n\n`;
  content += `_Generated at ${dates.time}_\n\n`;
  content += `</div>\n`;
  
  return content;
}

/**
 * Update README.md with the daily brief
 */
function updateReadme(briefContent) {
  try {
    let readme = fs.readFileSync(README_PATH, 'utf8');
    
    // Find and replace content between markers
    const startMarker = '<!-- BEGIN DAILY BRIEF -->';
    const endMarker = '<!-- END DAILY BRIEF -->';
    
    const startIndex = readme.indexOf(startMarker);
    const endIndex = readme.indexOf(endMarker);
    
    if (startIndex !== -1 && endIndex !== -1) {
      const before = readme.substring(0, startIndex + startMarker.length);
      const after = readme.substring(endIndex);
      
      const wrappedContent = `\n<details>\n<summary><b>📰 Today's Intelligence Brief</b></summary>\n\n${briefContent}\n\n</details>\n`;
      
      readme = before + wrappedContent + after;
      
      fs.writeFileSync(README_PATH, readme, 'utf8');
      console.log('✅ README.md updated successfully');
      return true;
    } else {
      console.error('⚠️  Daily brief markers not found in README.md');
      return false;
    }
  } catch (error) {
    console.error('❌ Error updating README:', error.message);
    return false;
  }
}

/**
 * Create daily archive file
 */
function createArchive(briefContent, dates) {
  try {
    const archivePath = path.join(DAILY_DIR, `${dates.iso}.md`);
    
    let archiveContent = `---\n`;
    archiveContent += `date: ${dates.iso}\n`;
    archiveContent += `title: "Daily Brief - ${dates.long}"\n`;
    archiveContent += `generated: ${new Date().toISOString()}\n`;
    archiveContent += `---\n\n`;
    archiveContent += briefContent;
    
    fs.writeFileSync(archivePath, archiveContent, 'utf8');
    console.log(`✅ Archive created: ${archivePath}`);
    return true;
  } catch (error) {
    console.error('❌ Error creating archive:', error.message);
    return false;
  }
}

/**
 * Main execution
 */
async function main() {
  console.log('🚀 Starting Daily Brief Generation\n');
  console.log('='.repeat(50) + '\n');
  
  // Ensure directories exist
  ensureDirectories();
  
  // Get date information
  const dates = getDateFormats();
  console.log(`📅 Date: ${dates.long}\n`);
  
  // Fetch all data
  const data = await fetchAllData();
  
  // Generate markdown content
  console.log('📝 Generating markdown content...');
  const briefContent = generateBriefMarkdown(data, dates);
  
  // Update README
  console.log('📄 Updating README.md...');
  const readmeUpdated = updateReadme(briefContent);
  
  // Create archive
  console.log('📁 Creating daily archive...');
  const archiveCreated = createArchive(briefContent, dates);
  
  // Summary
  console.log('\n' + '='.repeat(50));
  console.log('✨ Daily Brief Generation Complete!\n');
  console.log(`README Updated: ${readmeUpdated ? '✅' : '❌'}`);
  console.log(`Archive Created: ${archiveCreated ? '✅' : '❌'}`);
  console.log('\n' + '='.repeat(50));
  
  // Exit with appropriate code
  process.exit((readmeUpdated && archiveCreated) ? 0 : 1);
}

// Run if called directly
if (require.main === module) {
  main().catch(error => {
    console.error('❌ Fatal error:', error);
    process.exit(1);
  });
}

module.exports = { generateBriefMarkdown, fetchAllData };

File: scripts/generate_daily_brief.py
#!/usr/bin/env python3
"""
Daily Brief Orchestrator
Multi-agent architecture for generating the daily intelligence brief.
Spawns and coordinates independent agents, then assembles the final report.
"""

import sys
import json
import logging
from pathlib import Path
from datetime import datetime
from typing import Dict, List, Tuple
from concurrent.futures import ThreadPoolExecutor, as_completed

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger('Orchestrator')

# Add agents directory to path
sys.path.insert(0, str(Path(__file__).parent.parent / 'agents'))

# Import agents
from quote_agent import QuoteAgent
from weather_agent import WeatherAgent
from intel_agent import IntelAgent
from cyberpulse_agent import CyberPulseAgent
from trending_agent import TrendingAgent


class DailyBriefOrchestrator:
    """Orchestrates the multi-agent daily brief generation."""
    
    def __init__(self):
        self.root_dir = Path(__file__).parent.parent
        self.output_dir = self.root_dir / 'agents' / 'output'
        self.daily_dir = self.root_dir / 'daily'
        self.readme_path = self.root_dir / 'README.md'
        
        # Ensure directories exist
        self.output_dir.mkdir(parents=True, exist_ok=True)
        self.daily_dir.mkdir(parents=True, exist_ok=True)
        
        # Initialize agents
        self.agents = [
            QuoteAgent(),
            WeatherAgent(),
            IntelAgent(),
            CyberPulseAgent(),
            TrendingAgent(assets_dir=str(self.root_dir / 'assets'))
        ]
    
    def run_agents(self, parallel: bool = True) -> Dict[str, Tuple[bool, str]]:
        """
        Run all agents and collect their outputs.
        
        Args:
            parallel: Run agents in parallel (default: True)
            
        Returns:
            Dict mapping agent name to (success, output_path)
        """
        logger.info("=" * 60)
        logger.info("🚀 Starting Daily Brief Generation")
        logger.info("=" * 60)
        
        results = {}
        
        if parallel:
            # Run agents in parallel
            logger.info("Running agents in parallel...")
            with ThreadPoolExecutor(max_workers=5) as executor:
                future_to_agent = {executor.submit(agent.run): agent for agent in self.agents}
                
                for future in as_completed(future_to_agent):
                    agent = future_to_agent[future]
                    try:
                        success = future.result()
                        output_file = self.output_dir / f"{agent.name.lower()}.md"
                        results[agent.name] = (success, str(output_file))
                        
                        status = "✅" if success else "❌"
                        logger.info(f"{status} {agent.name}Agent: {'Success' if success else 'Failed'}")
                    except Exception as e:
                        logger.error(f"❌ {agent.name}Agent exception: {e}")
                        results[agent.name] = (False, None)
        else:
            # Run agents sequentially
            logger.info("Running agents sequentially...")
            for agent in self.agents:
                try:
                    success = agent.run()
                    output_file = self.output_dir / f"{agent.name.lower()}.md"
                    results[agent.name] = (success, str(output_file))
                    
                    status = "✅" if success else "❌"
                    logger.info(f"{status} {agent.name}Agent: {'Success' if success else 'Failed'}")
                except Exception as e:
                    logger.error(f"❌ {agent.name}Agent exception: {e}")
                    results[agent.name] = (False, None)
        
        return results
    
    def collect_outputs(self) -> Dict[str, str]:
        """
        Collect markdown outputs from all agents.
        
        Returns:
            Dict mapping agent name to markdown content
        """
        outputs = {}
        
        for agent in self.agents:
            md_file = self.output_dir / f"{agent.name.lower()}.md"
            
            if md_file.exists():
                outputs[agent.name] = md_file.read_text(encoding='utf-8')
                logger.info(f"📄 Collected output from {agent.name}Agent")
            else:
                outputs[agent.name] = f"**{agent.name} data unavailable** ⛔\n"
                logger.warning(f"⚠️  No output file for {agent.name}Agent")
        
        return outputs
    
    def generate_brief(self, outputs: Dict[str, str]) -> str:
        """
        Generate the unified daily brief markdown.
        
        Args:
            outputs: Dict of agent outputs
            
        Returns:
            Complete markdown brief
        """
        now = datetime.now()
        date_long = now.strftime('%A, %B %d, %Y')
        time_str = now.strftime('%I:%M %p %Z')
        
        markdown = f"<div align=\"center\">\n\n"
        markdown += f"# 📅 Daily Brief\n\n"
        markdown += f"**{date_long}**\n\n"
        markdown += f"</div>\n\n"
        markdown += f"---\n\n"
        
        # 1. Quote of the Day
        markdown += f"<details>\n<summary><b>💭 Quote of the Day</b></summary>\n\n"
        markdown += outputs.get('Quote', '**Quote unavailable** ⛔\n')
        markdown += f"\n</details>\n\n"
        
        # 2. Weather & Space Weather
        markdown += f"<details>\n<summary><b>🌤️ Weather Report</b></summary>\n\n"
        markdown += outputs.get('Weather', '**Weather data unavailable** ⛔\n')
        markdown += f"\n</details>\n\n"
        
        # 3. Global Intelligence Report
        markdown += f"<details>\n<summary><b>📰 Global Intelligence News</b></summary>\n\n"
        markdown += outputs.get('Intel', '**No news stories available** ⛔\n')
        markdown += f"\n</details>\n\n"
        
        # 4. Cyber Pulse Report
        markdown += f"<details>\n<summary><b>🔐 Cyber Pulse Report</b></summary>\n\n"
        markdown += outputs.get('CyberPulse', '**No cybersecurity headlines available** ⛔\n')
        markdown += f"\n</details>\n\n"
        
        # 5. Trending Repos
        markdown += f"<details>\n<summary><b>🔥 Trending on GitHub</b></summary>\n\n"
        markdown += outputs.get('Trending', '**No trending repositories available** ⛔\n')
        markdown += f"\n</details>\n\n"
        
        markdown += f"---\n\n"
        markdown += f"<div align=\"center\">\n\n"
        markdown += f"_Generated at {time_str}_\n\n"
        markdown += f"</div>\n"
        
        return markdown
    
    def update_readme(self, brief_content: str) -> bool:
        """
        Update README.md with the daily brief.
        
        Args:
            brief_content: The generated brief markdown
            
        Returns:
            True if successful, False otherwise
        """
        try:
            readme = self.readme_path.read_text(encoding='utf-8')
            
            # Find markers
            start_marker = '<!-- BEGIN DAILY BRIEF -->'
            end_marker = '<!-- END DAILY BRIEF -->'
            
            start_idx = readme.find(start_marker)
            end_idx = readme.find(end_marker)
            
            if start_idx == -1 or end_idx == -1:
                logger.error("⚠️  Daily brief markers not found in README.md")
                return False
            
            # Build new content
            before = readme[:start_idx + len(start_marker)]
            after = readme[end_idx:]
            
            wrapped_content = (
                f"\n<details>\n"
                f"<summary><b>📰 Today's Intelligence Brief</b></summary>\n\n"
                f"{brief_content}\n\n"
                f"</details>\n"
            )
            
            new_readme = before + wrapped_content + after
            
            # Write updated README
            self.readme_path.write_text(new_readme, encoding='utf-8')
            logger.info("✅ README.md updated successfully")
            return True
            
        except Exception as e:
            logger.error(f"❌ Error updating README: {e}")
            return False
    
    def create_archive(self, brief_content: str) -> bool:
        """
        Create daily archive file.
        
        Args:
            brief_content: The generated brief markdown
            
        Returns:
            True if successful, False otherwise
        """
        try:
            now = datetime.now()
            date_iso = now.strftime('%Y-%m-%d')
            date_long = now.strftime('%A, %B %d, %Y')
            
            archive_path = self.daily_dir / f"{date_iso}.md"
            
            archive_content = f"---\n"
            archive_content += f"date: {date_iso}\n"
            archive_content += f"title: \"Daily Brief - {date_long}\"\n"
            archive_content += f"generated: {now.isoformat()}\n"
            archive_content += f"---\n\n"
            archive_content += brief_content
            
            archive_path.write_text(archive_content, encoding='utf-8')
            logger.info(f"✅ Archive created: {archive_path}")
            return True
            
        except Exception as e:
            logger.error(f"❌ Error creating archive: {e}")
            return False
    
    def print_summary(self, results: Dict[str, Tuple[bool, str]]) -> None:
        """Print execution summary."""
        logger.info("")
        logger.info("=" * 60)
        logger.info("✨ Daily Brief Generation Complete!")
        logger.info("=" * 60)
        
        for agent_name, (success, _) in results.items():
            status = "✅" if success else "❌"
            logger.info(f"{status} {agent_name}Agent: {'Success' if success else 'Failed'}")
    
    def run(self) -> int:
        """
        Execute the complete daily brief generation workflow.
        
        Returns:
            Exit code (0 for success, 1 for failure)
        """
        try:
            # Step 1: Run all agents
            results = self.run_agents(parallel=True)
            
            # Step 2: Collect outputs
            logger.info("")
            logger.info("📦 Collecting agent outputs...")
            outputs = self.collect_outputs()
            
            # Step 3: Generate unified brief
            logger.info("📝 Generating unified daily brief...")
            brief_content = self.generate_brief(outputs)
            
            # Step 4: Update README
            logger.info("📄 Updating README.md...")
            readme_updated = self.update_readme(brief_content)
            
            # Step 5: Create archive
            logger.info("📁 Creating daily archive...")
            archive_created = self.create_archive(brief_content)
            
            # Print summary
            self.print_summary(results)
            
            logger.info("")
            logger.info(f"README Updated: {'✅' if readme_updated else '❌'}")
            logger.info(f"Archive Created: {'✅' if archive_created else '❌'}")
            
            return 0
            
        except Exception as e:
            logger.error(f"❌ Fatal error: {e}", exc_info=True)
            return 1


def main():
    """Main entry point."""
    orchestrator = DailyBriefOrchestrator()
    return orchestrator.run()


if __name__ == "__main__":
    sys.exit(main())
